types/src/error.rs

```diff
@@ -177,32 +177,44 @@ impl Error {
 	}
 }
 
-/// Error type with a special `subscription_closed` field to detect that
+/// A type with a special `subscription_closed` field to detect that
 /// a subscription has been closed to distinguish valid items produced
 /// by the server on the subscription stream from an error.
 #[derive(Deserialize, Serialize, Debug, PartialEq)]
-pub struct SubscriptionClosedError {
-	subscription_closed: String,
-	id: u64,
+pub struct SubscriptionClosed {
+	subscription_closed: SubscriptionClosedReason,
 }
 
-impl SubscriptionClosedError {
-	/// Create a new subscription closed error.
-	pub fn new(reason: impl Into<String>, id: u64) -> Self {
-		Self { subscription_closed: reason.into(), id }
+impl From<SubscriptionClosedReason> for SubscriptionClosed {
+	fn from(kind: SubscriptionClosedReason) -> Self {
+		Self::new(kind)
 	}
+}
 
-	/// Get the reason why the subscription was closed.
-	pub fn close_reason(&self) -> &str {
-		&self.subscription_closed
+impl SubscriptionClosed {
+	/// Create a new subscription message.
+	pub fn new(kind: SubscriptionClosedReason) -> Self {
+		Self { subscription_closed: kind }
 	}
 
-	/// Get the subscription ID.
-	pub fn subscription_id(&self) -> u64 {
-		self.id
+	/// Get the close reason.
+	pub fn close_reason(&self) -> &SubscriptionClosedReason {
+		&self.subscription_closed
 	}
 }
 
+/// A type to represent when a subscription gets closed
+/// by either the server or client side.
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
+pub enum SubscriptionClosedReason {
```

Comment:
I removed the subscription ID from here because `SubscriptionSink` inserts in the subid in the message anyway.

Reply:
That's fine; do we have a test somewhere that asserts that the ID is still available?

Reply:
we do test this in the `integration tests` and `RpcModule::subscribe` tests because it's required that the response is a `SubscriptionResponse` i.e, has a field `subscription` where the `sub id` is provided.

If that is missing such as "ordinary notification without subscription field` it would not be decoded by the client